File: product/wp10/commands.py
"""Owner-authorized local event commands, separate from the subscription port.

No keys or peer-supplied authorization are accepted over this surface. The trusted
embedding selects one already-open EventHost and grants publication per attached
private descriptor. The grant cannot exceed the journal's current write authority.
"""
from __future__ import annotations

import asyncio
from collections import deque
import copy
from dataclasses import dataclass
import hashlib
import re

from .events import EventError, decode, need, random_bytes, validate_payload
from .host import EventHost, MAX_REQUEST, plain
from .sdk import from_hex, record

PROTOCOL = 'par-local-event-commands-0040'
OPERATIONS = frozenset(('publish', 'inquire'))


def payload_from_wire(schema: dict, fields: dict) -> dict:
    """Validate canonical integer strings before converting to Python integers."""
    try:
        record(fields, schema)
        out = {}
        for name, kind in schema.items():
            field = fields[name]
            record(field, ('kind', 'value'))
            need(field['kind'] == kind, 'COMMAND_INPUT_INVALID')
            value = field['value']
            if kind in ('uint64', 'int64'):
                need(type(value) is str and len(value) <= 20 and
                     re.fullmatch(r'0|-?[1-9][0-9]*', value) is not None, 'COMMAND_INPUT_INVALID')
                number = int(value)
                need(0 <= number < 2**64 if kind == 'uint64' else -(2**63) <= number < 2**63,
                     'COMMAND_INPUT_INVALID')
                out[name] = number
            elif kind == 'bytes':
                need(type(value) is str and len(value) <= 32768 and len(value) % 2 == 0 and
                     re.fullmatch('[0-9a-f]*', value) is not None, 'COMMAND_INPUT_INVALID')
                out[name] = bytes.fromhex(value)
            else:
                need(type(value) is (str if kind == 'text' else bool), 'COMMAND_INPUT_INVALID')
                out[name] = value
        validate_payload(schema, out)  # Exact canonical encoded 16 KiB budget.
        return out
    except EventError as exc:
        if exc.code == 'RESOURCE_LIMIT': raise
        raise EventError('COMMAND_INPUT_INVALID') from None


@dataclass(eq=False)
class _Pending:
    channel: str
    operation: str
    args: dict
    size: int
    future: asyncio.Future
    released: bool = False


@dataclass
class _Attachment:
    context: dict
    allow_publish: bool
    pending: _Pending | None = None


class EventCommandHost:
    """Bounded cooperative command mailbox sharing the EventHost owner thread.

    At most one outstanding command per attachment; queue count/bytes reject
    synchronously. Cancellation of queued work prevents dispatch. Once synchronous
    SQLite/crypto starts, transport cancellation cannot preempt it. The journal's
    owner-side cancellation probe still observes cancellation at its defined fences.
    """
    def __init__(self, host: EventHost, *, max_channels=8, max_requests=16,
                 max_request_bytes=131072, turn_budget=2):
        host._enter()
        for value, low, high in ((max_channels, 1, 64), (max_requests, 1, 128),
                                 (max_request_bytes, 1024, 1048576), (turn_budget, 1, 16)):
            need(type(value) is int and low <= value <= high, 'HOST_OPTIONS_INVALID')
        self._host = host
        self._loop = asyncio.get_running_loop()
        self._max_channels, self._max_requests = max_channels, max_requests
        self._max_bytes, self._budget = max_request_bytes, turn_budget
        self._channels: dict[str, _Attachment] = {}
        self._queue: deque[_Pending] = deque()
        self._requests = self._bytes = self._peak_requests = self._peak_bytes = 0
        self._closed = False
        self._scheduled = None
        self._closed_event = asyncio.Event()
        self._watcher = self._loop.create_task(self._follow_owner())

    async def _follow_owner(self):
        try: await self._host.wait_closed()
        except asyncio.CancelledError: return
        self.close()

    def _enter(self):
        self._host._enter()
        need(not self._closed, 'HOST_CLOSED')

    def _channel(self, channel):
        need(type(channel) is str and channel in self._channels, 'HOST_CHANNEL_CLOSED')
        return self._channels[channel]

    def attach(self, *, allow_publish=False):
        self._enter()
        need(type(allow_publish) is bool, 'HOST_OPTIONS_INVALID')
        need(len(self._channels) < self._max_channels, 'HOST_CHANNEL_LIMIT')
        journal = self._host._j
        data = decode(journal._context)
        context = {'protocol': PROTOCOL, 'appId': data[1], 'spaceId': data[2].hex(),
                   'streamId': data[3].hex(), 'epoch': str(data[4]), 'schema': dict(journal._schema),
                   'schemaDigest': hashlib.sha256(data[5]).hexdigest(), 'issuer': journal._public.hex(),
                   'journalGeneration': journal._generation.hex(),
                   'authority': 'owner-publish' if allow_publish else 'inquire-only'}
        channel = random_bytes(16).hex()
        self._channels[channel] = _Attachment(context, allow_publish)
        return channel

    def context(self, channel):
        self._enter()
        return copy.deepcopy(self._channel(channel).context)

    def request(self, channel, operation, args):
        self._enter()
        need(type(operation) is str and operation in OPERATIONS, 'HOST_OPERATION_DENIED')
        attachment = self._channel(channel)
        need(attachment.pending is None, 'HOST_CHANNEL_BUSY')
        copied, size = plain(args, MAX_REQUEST)
        need(type(copied) is dict, 'COMMAND_INPUT_INVALID')
        need(self._requests < self._max_requests and self._bytes + size <= self._max_bytes, 'HOST_QUEUE_FULL')
        future = self._loop.create_future()
        pending = _Pending(channel, operation, copied, size, future)
        attachment.pending = pending
        self._queue.append(pending)
        self._requests += 1; self._bytes += size
        self._peak_requests = max(self._peak_requests, self._requests)
        self._peak_bytes = max(self._peak_bytes, self._bytes)
        future.add_done_callback(lambda f: self._cancel(pending) if f.cancelled() else None)
        self._schedule()
        return future

    def _schedule(self):
        if self._scheduled is None and not self._closed:
            self._scheduled = self._loop.call_soon(self._drain)

    def _release(self, pending):
        if pending.released: return
        pending.released = True
        self._requests -= 1; self._bytes -= pending.size
        attachment = self._channels.get(pending.channel)
        if attachment and attachment.pending is pending: attachment.pending = None
        pending.args = {}  # Drop plaintext references even when a done callback retains this record.

    def _cancel(self, pending):
        self._release(pending)
        try: self._queue.remove(pending)
        except ValueError: pass

    def _finish(self, pending, value=None, error=None):
        self._release(pending)
        if pending.future.done(): return
        if error is not None: pending.future.set_exception(error)
        else: pending.future.set_result(value)

    def _drain(self):
        self._scheduled = None
        for _ in range(self._budget):
            if self._closed or not self._queue: break
            pending = self._queue.popleft()
            if pending.released or pending.future.cancelled(): self._release(pending); continue
            try: self._finish(pending, self._execute(pending))
            except Exception as exc:
                code = exc.code if isinstance(exc, EventError) else 'HOST_OPERATION_FAILED'
                self._finish(pending, error=EventError(code))
        if self._queue: self._schedule()

    def _execute(self, pending):
        self._enter()
        attachment = self._channel(pending.channel)
        args = pending.args
        # An invalid/missing ID cannot be used as a correlation identity in a result.
        operation_id = from_hex(args.get('operationId'), 16)
        base = {'context': copy.deepcopy(attachment.context), 'hostId': self._host.stats()['hostId'],
                'operationId': operation_id.hex()}
        try:
            expected_keys = ('context', 'operationId', 'payload', 'parents') if pending.operation == 'publish' else ('context', 'operationId')
            record(args, expected_keys)
            need(type(args['context']) is dict and args['context'] == attachment.context, 'COMMAND_CONTEXT_MISMATCH')
            if pending.operation == 'inquire':
                receipt = self._host._j.inspect_operation(operation_id)
                if receipt is None: return base | {'kind': 'not-found-local'}
            else:
                need(attachment.allow_publish, 'COMMAND_NOT_AUTHORIZED')
                payload = payload_from_wire(self._host._j._schema, args['payload'])
                parents = args['parents']
                need(type(parents) is list and len(parents) <= 16, 'COMMAND_INPUT_INVALID')
                parents = [from_hex(p, 32) for p in parents]
                need(len(set(parents)) == len(parents), 'COMMAND_INPUT_INVALID')
                receipt = self._host.publish(operation_id, payload, parents=parents, cancelled=pending.future.cancelled)
            if receipt.state == 'CANCELLED':
                return base | {'kind': 'cancelled', 'phase': 'before-commit'}
            return base | {'kind': 'local-committed', 'sequence': str(receipt.sequence),
                           'eventId': receipt.event_id.hex(), 'replicated': False,
                           'cancellationRequested': receipt.cancellation_requested}
        except EventError as exc:
            if exc.code == 'EVENT_OUTCOME_UNKNOWN':
                return base | {'kind': 'outcome-unknown', 'code': 'LOCAL_OUTCOME_UNKNOWN'}
            # Journal errors other than EVENT_OUTCOME_UNKNOWN precede any uncertain
            # commit, or concern read-only inquiry. Never reinterpret transport errors.
            return base | {'kind': 'rejected', 'code': exc.code}

    def detach(self, channel):
        self._host._owner()
        attachment = self._channels.get(channel)
        if attachment is None: return
        if attachment.pending is not None:
            self._finish(attachment.pending, error=EventError('HOST_CHANNEL_CLOSED'))
        self._queue = deque(p for p in self._queue if p.channel != channel)
        del self._channels[channel]

    def stats(self):
        self._host._owner()
        return {'hostId': self._host.stats()['hostId'], 'closed': self._closed,
                'channels': len(self._channels), 'requests': self._requests, 'request_bytes': self._bytes,
                'peak_requests': self._peak_requests, 'peak_request_bytes': self._peak_bytes,
                'automatic_retry': False, 'automatic_ack': False}

    def close(self):
        self._host._owner()
        if self._closed: return
        for channel in list(self._channels): self.detach(channel)
        self._closed = True
        if self._scheduled: self._scheduled.cancel(); self._scheduled = None
        self._closed_event.set()
        if self._watcher is not asyncio.current_task(): self._watcher.cancel()

    async def wait_closed(self):
        self._host._owner()
        await self._closed_event.wait()
        if self._watcher is not asyncio.current_task():
            await asyncio.gather(self._watcher, return_exceptions=True)

File: product/wp10/src/commands.ts
/** Dedicated local commands. No shared-document commit, ACK or automatic replay. */
import {EventBindingError,copyData,obj,need,hex,u64,payload,same} from './contracts.js';
import type {Kind,Decimal} from './contracts.js';

export const COMMAND_PROTOCOL='par-local-event-commands-0040' as const;
export type CommandOperation='publish'|'inquire';
export interface CommandContext {
 readonly protocol:typeof COMMAND_PROTOCOL;
 readonly appId:string; readonly spaceId:string; readonly streamId:string;
 readonly epoch:Decimal; readonly schema:Readonly<Record<string,Kind>>;
 readonly schemaDigest:string; readonly issuer:string; readonly journalGeneration:string;
 readonly authority:'owner-publish'|'inquire-only';
}
export type CommandField =
 | Readonly<{kind:'text'|'uint64'|'int64'|'bytes';value:string}>
 | Readonly<{kind:'bool';value:boolean}>;
export interface PublishCommand {
 readonly operationId:string;
 readonly payload:Readonly<Record<string,CommandField>>;
 readonly parents:ReadonlyArray<string>;
}
export interface EventCommandChannel {
 request(operation:CommandOperation,args:unknown,signal:AbortSignal):Promise<unknown>;
}
export interface LocalEventCommit {
 readonly kind:'local-committed'; readonly operationId:string;
 readonly eventId:string; readonly sequence:Decimal;
 readonly replicated:false; readonly cancellationRequested:boolean;
}
export interface CommandCancelled {
 readonly kind:'cancelled'; readonly operationId:string; readonly phase:'before-send'|'before-commit';
}
export interface CommandRejected {
 readonly kind:'rejected'; readonly operationId:string; readonly code:string;
}
export interface LocalOutcomeUnknown {
 readonly kind:'outcome-unknown'; readonly operationId:string; readonly code:'LOCAL_OUTCOME_UNKNOWN';
}
export interface LocalEventAbsent {readonly kind:'not-found-local';readonly operationId:string;}
export type PublishResult=LocalEventCommit|CommandCancelled|CommandRejected|LocalOutcomeUnknown;
export type InquiryResult=LocalEventCommit|CommandCancelled|CommandRejected|LocalEventAbsent;
type CommandResult=PublishResult|InquiryResult;

export function commandContext(value:unknown):CommandContext {
 const r=obj(copyData(value,16384),['protocol','appId','spaceId','streamId','epoch','schema','schemaDigest','issuer','journalGeneration','authority']);
 need(r.protocol===COMMAND_PROTOCOL);
 need(typeof r.appId==='string'&&/^[a-z0-9][a-z0-9.-]{0,127}$/.test(r.appId));
 for(const k of ['spaceId','streamId','schemaDigest','issuer','journalGeneration'])hex(r[k],32);
 need(u64(r.epoch)>0n);need(r.authority==='owner-publish'||r.authority==='inquire-only');
 need(r.schema!==null&&typeof r.schema==='object'&&!Array.isArray(r.schema));
 const schema=r.schema as Record<string,unknown>,keys=Object.keys(schema);need(keys.length>0&&keys.length<=16);
 for(const key of keys)need(/^[a-z][a-z0-9_]{0,31}$/.test(key)&&['text','uint64','int64','bool','bytes'].includes(schema[key] as string));
 Object.freeze(schema);return Object.freeze(r) as unknown as CommandContext;
}
/** Exact bounded copy for both direct commands and retained client input.
 * This validates data, not caller authority or persistence across process restart.
 */
export function publishCommand(value:unknown,context:CommandContext):PublishCommand {
 try {
  const r=obj(copyData(value,65536),['operationId','payload','parents']);hex(r.operationId,16);
  payload(r.payload,context.schema);
  need(Array.isArray(r.parents)&&r.parents.length<=16);
  for(const p of r.parents)hex(p,32);
  need(new Set(r.parents).size===r.parents.length);
  need(new TextEncoder().encode(JSON.stringify({...r,context})).length<=65536);
  const fields=r.payload as Record<string,unknown>;
  for(const field of Object.values(fields))Object.freeze(field);
  Object.freeze(fields);Object.freeze(r.parents);
  return Object.freeze(r) as unknown as PublishCommand;
 } catch {throw new EventBindingError('COMMAND_INPUT_INVALID')}
}
const rejected=(operationId:string,code:string):CommandRejected=>Object.freeze({kind:'rejected',operationId,code});
const unknown=(operationId:string):LocalOutcomeUnknown=>Object.freeze({kind:'outcome-unknown',operationId,code:'LOCAL_OUTCOME_UNKNOWN'});

/** An owner-injected port, or ConnectedCommandChannel over a private descriptor.
 * A failed/unknown exchange is terminal for this client: rebind explicitly before
 * inquiry/retry. Read-only not-found-local is an observation, never replay consent.
 * Programmer input errors reject with EventBindingError(COMMAND_INPUT_INVALID).
 * Operational outcomes are discriminated unions; none of them automatically ACK.
 */
export class EventCommands {
 readonly #channel:EventCommandChannel;
 readonly #context:CommandContext;
 readonly #hostId:string;
 readonly #timeout:number;
 #busy=false; #closed=false; #failed=false; #active:AbortController|undefined;
 constructor(channel:EventCommandChannel,expectedContext:CommandContext,hostId:string,
             options:{requestTimeoutMs?:number}={}) {
  need(typeof channel?.request==='function','INVALID_PORT');
  this.#channel=channel;this.#context=commandContext(expectedContext);this.#hostId=hex(hostId,16);
  const opts=copyData(options,1024) as Record<string,unknown>;
  obj(opts,Object.hasOwn(opts,'requestTimeoutMs')?['requestTimeoutMs']:[]);
  const timeout=opts.requestTimeoutMs??6000;
  need(typeof timeout==='number'&&Number.isInteger(timeout)&&timeout>=10&&timeout<=60000,'INVALID_OPTIONS');
  this.#timeout=timeout;
 }
 async publish(input:PublishCommand,options:{signal?:AbortSignal}={}):Promise<PublishResult> {
  const args={...publishCommand(input,this.#context),context:this.#context};
  const operationId=args.operationId as string;
  if(this.#context.authority!=='owner-publish')return rejected(operationId,'COMMAND_NOT_AUTHORIZED');
  return await this.#perform('publish',args,options.signal) as PublishResult;
 }
 async inquire(operationId:string,options:{signal?:AbortSignal}={}):Promise<InquiryResult> {
  try {hex(operationId,16)} catch {throw new EventBindingError('COMMAND_INPUT_INVALID')}
  return await this.#perform('inquire',{context:this.#context,operationId},options.signal) as InquiryResult;
 }
 async #perform(operation:CommandOperation,args:Record<string,unknown>,signal:AbortSignal|undefined):Promise<CommandResult> {
  const operationId=args.operationId as string;
  if(signal!==undefined)need(typeof signal.aborted==='boolean'&&typeof signal.addEventListener==='function'&&typeof signal.removeEventListener==='function','INVALID_OPTIONS');
  if(this.#closed)return rejected(operationId,'COMMAND_CLOSED');
  if(this.#failed)return rejected(operationId,'COMMAND_RECONCILIATION_REQUIRED');
  if(this.#busy)return rejected(operationId,'COMMAND_BUSY');
  if(signal?.aborted)return Object.freeze({kind:'cancelled',operationId,phase:'before-send'});
  this.#busy=true;
  const controller=new AbortController();this.#active=controller;
  const relay=()=>controller.abort();signal?.addEventListener('abort',relay,{once:true});
  let abort:()=>void=()=>{};
  const interrupted=new Promise<never>((_,reject)=>{abort=()=>reject(new EventBindingError('COMMAND_INTERRUPTED'));controller.signal.addEventListener('abort',abort,{once:true})});
  const timer=setTimeout(relay,this.#timeout);
  try {
   // Admission is synchronous from the client's perspective. From here any thrown
   // error could follow a native commit; don't guess from an exception's message.
   // A synchronous throw must not leave the abort Promise unobserved.
   // This handler consumes only rejection; the race still receives it.
   void interrupted.catch(()=>{});
   const request=this.#channel.request(operation,args,controller.signal);
   const value=await Promise.race([request,interrupted]);
   need(!controller.signal.aborted,'COMMAND_INTERRUPTED');
   const result=this.#response(operation,operationId,value);
   if(result.kind==='outcome-unknown')this.#failed=true;
   return result;
  } catch {
   this.#failed=true;controller.abort();
   return operation==='publish'?unknown(operationId):rejected(operationId,'INQUIRY_UNAVAILABLE');
  } finally {
   clearTimeout(timer);controller.signal.removeEventListener('abort',abort);
   signal?.removeEventListener('abort',relay);this.#active=undefined;this.#busy=false;
  }
 }
 #response(operation:CommandOperation,operationId:string,value:unknown):CommandResult {
  const r=copyData(value,65536) as Record<string,unknown>;
  need(r&&typeof r==='object'&&!Array.isArray(r));
  need(r.hostId===this.#hostId&&r.operationId===operationId&&same(commandContext(r.context),this.#context));
  const base=['context','hostId','operationId','kind'];
  switch(r.kind){
   case 'local-committed':
    obj(r,[...base,'sequence','eventId','replicated','cancellationRequested']);
    need(u64(r.sequence)>0n);hex(r.eventId,32);need(r.replicated===false&&typeof r.cancellationRequested==='boolean');
    return Object.freeze({kind:'local-committed',operationId,sequence:r.sequence as string,eventId:r.eventId as string,replicated:false,cancellationRequested:r.cancellationRequested});
   case 'rejected':
    obj(r,[...base,'code']);need(typeof r.code==='string'&&/^[A-Z][A-Z0-9_]{0,63}$/.test(r.code));
    need(!['LOCAL_OUTCOME_UNKNOWN','EVENT_OUTCOME_UNKNOWN'].includes(r.code));
    return rejected(operationId,r.code);
   case 'cancelled':
    obj(r,[...base,'phase']);need(operation==='publish'&&r.phase==='before-commit');
    return Object.freeze({kind:'cancelled',operationId,phase:'before-commit'});
   case 'outcome-unknown':
    obj(r,[...base,'code']);need(operation==='publish'&&r.code==='LOCAL_OUTCOME_UNKNOWN');
    return unknown(operationId);
   case 'not-found-local':
    obj(r,base);need(operation==='inquire');return Object.freeze({kind:'not-found-local',operationId});
   default:throw new EventBindingError('COMMAND_RESPONSE_INVALID');
  }
 }
 /** Stop this client; the embedding still owns and explicitly closes its channel. */
 close():void {this.#closed=true;this.#active?.abort();}
 stats():Readonly<{closed:boolean;busy:boolean;reconciliationRequired:boolean}> {
  return Object.freeze({closed:this.#closed,busy:this.#busy,reconciliationRequired:this.#failed});
 }
}
